mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp

```diff
@@ -5330,6 +5423,9 @@ struct FoldTensorCastUnPackOp : public OpRewritePattern<UnPackOp> {
     if (!tensor::hasFoldableTensorCastOperand(op))
       return failure();
 
+    if (!op.hasPureTensorSemantics())
+      return failure();
```

Comment:
Let's add a TODO for consistency. It is a reasonable folder to me and we should support it (in follow-ups).